File: app/src/main/java/com/kakeibo/util/UtilFiles.java
package com.kakeibo.util;

import android.content.Context;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.ObjectOutputStream;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import static android.content.Context.MODE_PRIVATE;

public class UtilFiles {
    private static final String TAG = UtilFiles.class.getSimpleName();

    public static boolean fileExists(Context context, String filename) {
        File file = context.getFileStreamPath(filename);
        if(file == null || !file.exists()) {
            return false;
        }
        return true;
    }

    public static String getFileValue(String fileName, Context context) {
        try {
            StringBuilder stringBuilder = new StringBuilder();
            String inputLine = "";
            /*
             * We have to use the openFileInput()-method the ActivityContext
             * provides. Again for security reasons with openFileInput(...)
             */
            FileInputStream fIn = context.openFileInput(fileName);
            InputStreamReader isr = new InputStreamReader(fIn);
            BufferedReader inBuff = new BufferedReader(isr);
            while ((inputLine = inBuff.readLine()) != null) {
                stringBuilder.append(inputLine);
                stringBuilder.append("\n");
            }
            inBuff.close();
            return stringBuilder.toString();
        } catch (IOException e) {
            return null;
        }
    }

    public static ArrayList<String> getFileValues(String fileName, Context context) {
        try {
            StringBuilder sb = new StringBuilder();
            String inputLine = "";
            FileInputStream fIn = context.openFileInput(fileName);
            InputStreamReader isr = new InputStreamReader(fIn);
            BufferedReader inBuff = new BufferedReader(isr);
            while ((inputLine = inBuff.readLine()) != null) {
                sb.append(inputLine);
                sb.append("\n");
            }
            inBuff.close();
            return new ArrayList<>(Arrays.asList(sb.toString().split("\t")));
        } catch (IOException e) {
            return null;
        }
    }

    public static boolean appendFileValue(String fileName, String value, Context context) {
        return writeToFile(fileName, value, context, Context.MODE_APPEND);
    }

    public static boolean setFileValue(String fileName, String value, Context context) {
        return writeToFile(fileName, value, context, Context.MODE_WORLD_READABLE);
    }

    public static boolean writeToFile(String fileName, String value, Context context,
                                      int writeOrAppendMode) {
        // just make sure it's one of the modes we support
        // Context.MODE_PRIVATE: erase existing file
        // Context.MODE_APPEND: appends to an existing file
        if (writeOrAppendMode != Context.MODE_APPEND && writeOrAppendMode != MODE_PRIVATE) {
            return false;
        }

        //Log.d(TAG, "path="+context.getFilesDir().getAbsolutePath());

        try {
            /*
             * We have to use the openFileOutput()-method the ActivityContext
             * provides, to protect your file from others and This is done for
             * security-reasons. We chose MODE_WORLD_READABLE, because we have
             * nothing to hide in our file
             */
            FileOutputStream fOut = context.openFileOutput(fileName, writeOrAppendMode);
            OutputStreamWriter osw = new OutputStreamWriter(fOut);
            // Write the string to the file
            osw.write(value);
            // save and close
            osw.flush();
            osw.close();
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public static void deleteFile(String fileName, Context context) {
        context.deleteFile(fileName);
    }
}

File: app/src/main/java/com/kakeibo/util/UtilFiles.kt
package com.kakeibo.util

import android.content.Context
import com.kakeibo.core.data.local.entities.CategoryEntity
import com.kakeibo.core.data.local.entities.ItemEntity
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

object UtilFiles {
    val FILE_NAME = "to_be_exported"

    fun fileExists(context: Context, filename: String?): Boolean {
        val file = context.getFileStreamPath(filename)
        return !(file == null || !file.exists())
    }

    fun getFileValue(fileName: String, context: Context): String? {
        return try {
            val stringBuilder = StringBuilder()
            var inputLine: String?
            /*
             * We have to use the openFileInput()-method the ActivityContext
             * provides. Again for security reasons with openFileInput(...)
             */
            val fIn = context.openFileInput(fileName)
            val isr = InputStreamReader(fIn)
            val inBuff = BufferedReader(isr)
            while (inBuff.readLine().also { inputLine = it } != null) {
                stringBuilder.append(inputLine)
                stringBuilder.append("\n")
            }
            inBuff.close()
            stringBuilder.toString()
        } catch (e: IOException) {
            null
        }
    }

    fun getFileValues(fileName: String?, context: Context): ArrayList<String>? {
        return try {
            val sb = StringBuilder()
            var inputLine: String?
            val fIn = context.openFileInput(fileName)
            val isr = InputStreamReader(fIn)
            val inBuff = BufferedReader(isr)
            while (inBuff.readLine().also { inputLine = it } != null) {
                sb.append(inputLine)
                sb.append("\n")
            }
            inBuff.close()
            ArrayList(listOf(*sb.toString().split("\t").toTypedArray()))
        } catch (e: IOException) {
            null
        }
    }

    fun appendFileValue(fileName: String, value: String, context: Context): Boolean {
        return writeToFile(fileName, value, context, Context.MODE_APPEND)
    }

    fun setFileValue(fileName: String, value: String, context: Context): Boolean {
        return writeToFile(fileName, value, context, Context.MODE_WORLD_READABLE)
    }

    fun writeToFile(fileName: String, value: String, context: Context,
                    writeOrAppendMode: Int): Boolean {
        // just make sure it's one of the modes we support
        // Context.MODE_PRIVATE: erase existing file
        // Context.MODE_APPEND: appends to an existing file
        if (writeOrAppendMode != Context.MODE_APPEND && writeOrAppendMode != Context.MODE_PRIVATE) {
            return false
        }

        //Log.d(TAG, "path="+context.getFilesDir().getAbsolutePath());
        try {
            /*
             * We have to use the openFileOutput()-method the ActivityContext
             * provides, to protect your file from others and This is done for
             * security-reasons. We chose MODE_WORLD_READABLE, because we have
             * nothing to hide in our file
             */
            val fOut = context.openFileOutput(fileName, writeOrAppendMode)
            val osw = OutputStreamWriter(fOut)
            // Write the string to the file
            osw.write(value)
            // save and close
            osw.flush()
            osw.close()
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        }
        return true
    }

    fun deleteFile(fileName: String?, context: Context) {
        context.deleteFile(fileName)
    }

    /* Used in Import functionality */
    fun readCsvToList(
        list: MutableList<ItemEntity>,
        reader: BufferedReader,
        allCategoriesMap: Map<Int, CategoryEntity>
    ): String {
        var counter = 0
        val stringBuilder = StringBuilder()
        /*
         0: ordered by Date
            Event Date,Amount,Category,Memo,Updated Date,Category Code
         1: ordered by Category
            Category,Amount,Memo,Event Date,Updated Date,Category Code
         */
        reader.lineSequence().forEach {
            if (counter == 0) {
                // skip the header. it's written in different languages
            } else {
                val line = it.split(",")

                if (!UtilDate.isYMDDateValid(line[0])) {
                    stringBuilder.append("EventDate is not valid at line $counter \n")
                }
                if (!isAmountValid(line[1])) {
                    stringBuilder.append("Amount is not valid at line $counter \n")
                }
                if (!UtilCategory.isCategoryValid(line[5], line[2], allCategoriesMap)) {
                    stringBuilder.append("Category is not valid at line $counter \n")
                }
                if (!UtilDate.isYMDHSDateValid(line[4])) {
                    stringBuilder.append("UpdatedDate is not valid at line $counter \n")
                }

//                list.add(
//                    ItemEntity(
//                        BigDecimal(line[1].trim()), //amount
//                        "",
//                        line[5].trim().toInt(), // category code
//                        line[3].trim(), // memo
//                        line[0].trim(), // event date
//                        line[4].trim()  // updated date
//                    )
//                )
            }

            counter++
        }

        return stringBuilder.toString()
    }

    private fun isAmountValid(amount: String): Boolean {
        if (amount.toIntOrNull() == null)
            return false

        if (10 < amount.length)
            return false

        return true
    }
}